shell/platform/linux/fl_view.cc

```diff
@@ -255,12 +271,135 @@ static void fl_view_size_allocate(GtkWidget* widget,
   FlutterEngineSendWindowMetricsEvent(self->flutter_engine, &event);
 }
 
+static gboolean fl_view_button_press_event(GtkWidget* widget,
+                                           GdkEventButton* event) {
+  FlView* self = FL_VIEW(widget);
+
+  if (self->flutter_engine == nullptr)
+    return FALSE;
+
+  // Flutter doesn't handle double and triple click events
+  if (event->type == GDK_DOUBLE_BUTTON_PRESS ||
+      event->type == GDK_TRIPLE_BUTTON_PRESS)
+    return FALSE;
+
+  int64_t button = gdk_button_to_flutter_button(event->button);
+  if (button < 0)
+    return FALSE;
+
+  // Drop the event if Flutter already thinks the button is down
+  if ((self->button_state & button) != 0)
+    return FALSE;
+  self->button_state ^= button;
+
+  FlutterPointerEvent fl_event = {};
+  fl_event.struct_size = sizeof(fl_event);
+  fl_event.phase = kDown;
+  fl_event.x = event->x;
+  fl_event.y = event->y;
+  fl_event.device_kind = kFlutterPointerDeviceKindMouse;
+  fl_event.buttons = self->button_state;
+  FlutterEngineSendPointerEvent(self->flutter_engine, &fl_event, 1);
+
+  return TRUE;
+}
+
+static gboolean fl_view_button_release_event(GtkWidget* widget,
+                                             GdkEventButton* event) {
+  FlView* self = FL_VIEW(widget);
+
+  if (self->flutter_engine == nullptr)
+    return FALSE;
+
+  int64_t button = gdk_button_to_flutter_button(event->button);
+  if (button < 0)
+    return FALSE;
+
+  // Drop the event if Flutter already thinks the button is up
+  if ((self->button_state & button) == 0)
+    return FALSE;
+  self->button_state ^= button;
+
+  FlutterPointerEvent fl_event = {};
+  fl_event.struct_size = sizeof(fl_event);
+  fl_event.phase = kUp;
+  fl_event.x = event->x;
+  fl_event.y = event->y;
+  fl_event.device_kind = kFlutterPointerDeviceKindMouse;
+  fl_event.buttons = self->button_state;
+  FlutterEngineSendPointerEvent(self->flutter_engine, &fl_event, 1);
```

Comment:
There's a lot of duplicate code across these methods; it would be helpful to factor out a helper, as with the GLFW embedding's `SendPointerEventWithData`

Reply:
Yep, working on that, wanted to get something pointer related working before landing those changes. Since the keyboard handling is more complex I'll probably propose refactoring the engine code out and add some suitable methods to do this.

Reply:
I'm not sure I follow; keyboard events don't use the same struct or method, so I'm not clear how it would be related to making a pointer event helper.

Reply:
I mean that SendPointerEvent with data might become a method on the FlEngine object. But no specific plans here, just if it makes sense like that. I've moved all the logic into one method in this PR.